src/stream/src/executor/managed_state/aggregation/extreme.rs

```diff
@@ -123,261 +97,160 @@ pub trait ManagedTableState<S: StateStore>: Send + Sync + 'static {
+    // So we may introduce a trait for this later, and other code can be reused for each
+    // type of state.
+    fn get_output_from_cache(&self) -> Option<Datum> {
+        let row = self.cache.first()?;
+        Some(row[self.cache_agg_col_idx].clone())
     }
 
     async fn get_output_inner(
         &mut self,
         epoch: u64,
         state_table: &RowBasedStateTable<S>,
     ) -> StreamExecutorResult<Datum> {
-        // To make things easier, we do not allow get_output before flushing. Otherwise we will need
-        // to merge data from flush_buffer and state store, which is hard to implement.
-        //
-        // To make ExtremeState produce the correct result, the write batch must be flushed into the
-        // state store before getting the output. Note that calling `.flush()` is not enough, as it
-        // only generates a write batch without flushing to store.
-
-        // Firstly, check if datum is available in cache.
-        if let Some(v) = self.get_output_from_cache() {
-            Ok(Some(v))
+        // try to get the result from cache
+        if let Some(datum) = self.get_output_from_cache() {
+            Ok(datum)
         } else {
-            // Then, fetch from the state store.
-            //
-            // To future developers: please make **SURE** you have taken `EXTREME_TYPE` into
-            // account. EXTREME_MIN and EXTREME_MAX will significantly impact the
-            // following logic.
+            // read from state table and fill in the cache
             let all_data_iter =
                 iter_state_table(state_table, epoch, self.group_key.as_ref()).await?;
             pin_mut!(all_data_iter);
 
-            for _ in 0..self.top_n_count.unwrap_or(usize::MAX) {
-                if let Some(inner) = all_data_iter.next().await {
-                    let row = inner?;
-
-                    let group_key_len = self.group_key.as_ref().map_or(0, |row| row.size());
-                    // Get the agg call value. Same as sort key.
-                    let value = row[group_key_len].clone();
-
-                    // Get sort key and extreme pk.
-                    let sort_key = value.as_ref().map(|v| v.clone().try_into().unwrap());
-                    let mut extreme_pk = ExtremePk::with_capacity(1);
-                    // The layout is group_key/sort_key/extreme_pk. So the range
-                    // should be [group_key_len + 1, row.0.len()).
-                    for extreme_pk_index in group_key_len + 1..row.0.len() {
-                        extreme_pk.push(row[extreme_pk_index].clone());
-                    }
-
-                    self.top_n.insert((sort_key, extreme_pk), value);
-                } else {
-                    break;
-                }
+            self.cache.set_synced(); // after the following loop the cache should be synced
+
+            #[for_await]
+            for state_row in all_data_iter.take(self.cache.capacity.unwrap_or(usize::MAX)) {
```

Comment:
Actually `None` capacity means "no cache" instead of "full cache" before... To avoid misusing it anymore, I suggest just removing this `Option` and let the caller specify `0` or `usize::MAX` for both cases.